### DISCUSSION TITLE
Constants as array keys

### DISCUSSION THREAD

**Discussion by jrcii** [vimeo/psalm#9190] [Psalm development (AKA internals)]
A few years ago @muglug stated that (paraphrasing) "[array keys can't be indirect"](https://github.com/vimeo/psalm/issues/3399#issuecomment-631006942) and therefore can't be constants.

If there exists a string used both in code and would-appear-if-possible in various type definitions as keys, how would you DRY that if this isn't possible?

Is there a set/type theory/design reason for this limitation, or, is it just a hard implementation problem either in general or due to Psalm's internal architecture?

**Comment by orklah** [comment]
Not sure if it was a Parser limitation. I can't think of any other reason. If you can get that to work, I'd merge it :)
